src/libraries/QuoterMath.sol

```diff
@@ -0,0 +1,203 @@
+        );
+        initializedTicksCrossed = 1;
+
+        Slot0Struct memory slot0 = fillSlot0(poolManager, poolKey);
+
+        SwapState memory state = SwapState({
+            amountSpecifiedRemaining: -swapParams.amountSpecified,
+            amountCalculated: 0,
+            sqrtPriceX96: slot0.sqrtPriceX96,
+            tick: slot0.tick,
+            feeGrowthGlobalX128: 0,
+            protocolFee: 0,
+            liquidity: poolManager.getLiquidity(poolKey.toId())
+        });
+
+        while (state.amountSpecifiedRemaining != 0 && state.sqrtPriceX96 != quoteParams.sqrtPriceLimitX96) {
+            StepComputations memory step;
+
+            step.sqrtPriceStartX96 = state.sqrtPriceX96;
+
+            (step.tickNext, step.initialized) = PoolTickBitmap.nextInitializedTickWithinOneWord(
+                poolManager, poolKey.toId(), slot0.tickSpacing, state.tick, quoteParams.zeroForOne
+            );
+
+            // ensure that we do not overshoot the min/max tick, as the tick bitmap is not aware of these bounds
+            if (step.tickNext < TickMath.MIN_TICK) {
+                step.tickNext = TickMath.MIN_TICK;
+            } else if (step.tickNext > TickMath.MAX_TICK) {
+                step.tickNext = TickMath.MAX_TICK;
+            }
+
+            // get the price for the next tick
+            step.sqrtPriceNextX96 = TickMath.getSqrtPriceAtTick(step.tickNext);
+
+            // compute values to swap to the target tick, price limit, or point where input/output amount is exhausted
+            (state.sqrtPriceX96, step.amountIn, step.amountOut, step.feeAmount) = SwapMath.computeSwapStep(
+                state.sqrtPriceX96,
+                (
+                    quoteParams.zeroForOne
+                        ? step.sqrtPriceNextX96 < quoteParams.sqrtPriceLimitX96
+                        : step.sqrtPriceNextX96 > quoteParams.sqrtPriceLimitX96
+                ) ? quoteParams.sqrtPriceLimitX96 : step.sqrtPriceNextX96,
+                state.liquidity,
+                -state.amountSpecifiedRemaining,
+                quoteParams.fee
+            );
+
+            if (quoteParams.exactInput) {
+                state.amountSpecifiedRemaining -= (step.amountIn + step.feeAmount).toInt256();
+                state.amountCalculated = state.amountCalculated + step.amountOut.toInt256();
+            } else {
+                state.amountSpecifiedRemaining += step.amountOut.toInt256();
+                state.amountCalculated = state.amountCalculated - (step.amountIn + step.feeAmount).toInt256();
+            }
+
+            // shift tick if we reached the next price
+            if (state.sqrtPriceX96 == step.sqrtPriceNextX96) {
+                // if the tick is initialized, run the tick transition
+                if (step.initialized) {
+                    (, int128 liquidityNet,,) = poolManager.getTickInfo(poolKey.toId(), step.tickNext);
```

Comment:
in case ViewQuoter is used onchain, you can make this slightly more efficient with `poolManager.getTickLiquidity` as it will only read 1 slot (vs 3 slots)

Reply:
ooh didn't see that function thanks